packages/firebase_auth/firebase_auth_platform_interface/test/method_channel_firebase_auth_test.dart

```diff
@@ -0,0 +1,1238 @@
+      ]);
+    });
+
+    test('PhoneAuthProvider unlinkFromProvider', () async {
+      const PhoneAuthCredential phoneCredential = PhoneAuthCredential(
+        verificationId: kMockVerificationId,
+        smsCode: kMockSmsCode,
+      );
+      await auth.unlinkFromProvider(appName, phoneCredential.providerId);
+      expect(log, <Matcher>[
+        isMethodCall(
+          'unlinkFromProvider',
+          arguments: <String, String>{
+            'app': appName,
+            'provider': 'phone',
+          },
+        ),
+      ]);
+    });
+
+    test('TwitterAuthProvider unlinkFromProvider', () async {
+      const TwitterAuthCredential twitterCredential = TwitterAuthCredential(
+        authToken: kMockAuthToken,
+        authTokenSecret: kMockAuthTokenSecret,
+      );
+      await auth.unlinkFromProvider(appName, twitterCredential.providerId);
+      expect(log, <Matcher>[
+        isMethodCall(
+          'unlinkFromProvider',
+          arguments: <String, String>{
+            'app': appName,
+            'provider': 'twitter.com',
+          },
+        ),
+      ]);
+    });
+
+    test('GithubAuthProvider unlinkFromProvider', () async {
+      const GithubAuthCredential githubCredential =
+          GithubAuthCredential(token: kMockGithubToken);
+      await auth.unlinkFromProvider(appName, githubCredential.providerId);
+      expect(log, <Matcher>[
+        isMethodCall(
+          'unlinkFromProvider',
+          arguments: <String, String>{
+            'app': appName,
+            'provider': 'github.com',
+          },
+        ),
+      ]);
+    });
+
+    test('signInWithCustomToken', () async {
+      final PlatformAuthResult result =
+          await auth.signInWithCustomToken(appName, kMockCustomToken);
+      verifyAuthResult(result);
+      expect(
+        log,
+        <Matcher>[
+          isMethodCall('signInWithCustomToken', arguments: <String, String>{
+            'token': kMockCustomToken,
+            'app': appName,
+          })
+        ],
+      );
+    });
+
+    test('onAuthStateChanged', () async {
+      mockHandleId = 42;
+
+      Future<void> simulateEvent(Map<String, dynamic> user) async {
+        // TODO(hterkelsen): Remove this when defaultBinaryMessages is in stable.
+        // https://github.com/flutter/flutter/issues/33446
+        // ignore: deprecated_member_use
+        await BinaryMessages.handlePlatformMessage(
+          MethodChannelFirebaseAuth.channel.name,
+          MethodChannelFirebaseAuth.channel.codec.encodeMethodCall(
+            MethodCall(
+              'onAuthStateChanged',
+              <String, dynamic>{'id': 42, 'user': user, 'app': appName},
+            ),
+          ),
+          (_) {},
+        );
+      }
+
+      final AsyncQueue<PlatformUser> events = AsyncQueue<PlatformUser>();
+
+      // Subscribe and allow subscription to complete.
+      final StreamSubscription<PlatformUser> subscription =
+          auth.onAuthStateChanged(appName).listen(events.add);
```

Comment:
I wonder if we could use the `Stream` APIs to do this without a `AsyncQueue` and zero-duration await. Would this work?

```
Stream stream = auth.onAuthStateChanged(appName);
await simulateEvent(null);
expect(await stream.take(1).first, isNull);
await simulateEvent(kMockUser);
verifyUser(await stream.take(1).first);
```

Reply:
The test is hanging when I make that change. I think it's because the returned stream is a broadcast stream, so when you do `stream.take(1).first`, the event has already been fired and it waits forever.

Reply:
OK it's fine as-is then.